authority/admin/collection.go

```diff
@@ -0,0 +1,180 @@
+package admin
+
+import (
+	"crypto/sha1"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/smallstep/certificates/authority/provisioner"
+	"github.com/smallstep/certificates/linkedca"
+)
+
+// DefaultAdminLimit is the default limit for listing provisioners.
+const DefaultAdminLimit = 20
+
+// DefaultAdminMax is the maximum limit for listing provisioners.
+const DefaultAdminMax = 100
+
+type uidAdmin struct {
+	admin *linkedca.Admin
+	uid   string
+}
+
+type adminSlice []uidAdmin
+
+func (p adminSlice) Len() int           { return len(p) }
+func (p adminSlice) Less(i, j int) bool { return p[i].uid < p[j].uid }
+func (p adminSlice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+// Collection is a memory map of admins.
+type Collection struct {
+	byID                    *sync.Map
+	bySubProv               *sync.Map
+	byProv                  *sync.Map
+	sorted                  adminSlice
+	provisioners            *provisioner.Collection
+	superCount              int
+	superCountByProvisioner map[string]int
+}
+
+// NewCollection initializes a collection of provisioners. The given list of
+// audiences are the audiences used by the JWT provisioner.
+func NewCollection(provisioners *provisioner.Collection) *Collection {
+	return &Collection{
+		byID:                    new(sync.Map),
+		byProv:                  new(sync.Map),
+		bySubProv:               new(sync.Map),
+		superCountByProvisioner: map[string]int{},
+		provisioners:            provisioners,
+	}
+}
+
+// LoadByID a admin by the ID.
+func (c *Collection) LoadByID(id string) (*linkedca.Admin, bool) {
+	return loadAdmin(c.byID, id)
+}
+
+func subProvNameHash(sub, provName string) string {
+	subHash := sha1.Sum([]byte(sub))
+	provNameHash := sha1.Sum([]byte(provName))
+	_res := sha1.Sum(append(subHash[:], provNameHash[:]...))
+	return string(_res[:])
+}
```

Comment:
In go you can also use structs as keys, anything that is comparable. You should not use pointers, because then you will be comparing a memory address:

```go
type Foo struct {
    A string
    B string
}

Foo{"foo","bar"} == Foo{"foo","bar"} // true
Foo{"foo","bar"} == Foo{"bar","bar"} // false
&Foo{"foo","bar"} == &Foo{"foo","bar"} // false
```